// @azure/search-documents/dist-esm/src/generated/service/operations/indexes.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as coreHttp from "@azure/core-http";
import * as Mappers from "../models/indexesMappers";
import * as Parameters from "../models/parameters";
/** Class representing a Indexes. */
export class Indexes {
    /**
     * Create a Indexes.
     * @param {SearchServiceClientContext} client Reference to the service client.
     */
    constructor(client) {
        this.client = client;
    }
    create(index, options, callback) {
        return this.client.sendOperationRequest({
            index,
            options
        }, createOperationSpec, callback);
    }
    list(options, callback) {
        return this.client.sendOperationRequest({
            options
        }, listOperationSpec, callback);
    }
    createOrUpdate(indexName, index, options, callback) {
        return this.client.sendOperationRequest({
            indexName,
            index,
            options
        }, createOrUpdateOperationSpec, callback);
    }
    deleteMethod(indexName, options, callback) {
        return this.client.sendOperationRequest({
            indexName,
            options
        }, deleteMethodOperationSpec, callback);
    }
    get(indexName, options, callback) {
        return this.client.sendOperationRequest({
            indexName,
            options
        }, getOperationSpec, callback);
    }
    getStatistics(indexName, options, callback) {
        return this.client.sendOperationRequest({
            indexName,
            options
        }, getStatisticsOperationSpec, callback);
    }
    analyze(indexName, request, options, callback) {
        return this.client.sendOperationRequest({
            indexName,
            request,
            options
        }, analyzeOperationSpec, callback);
    }
}
// Operation Specifications
const serializer = new coreHttp.Serializer(Mappers);
const createOperationSpec = {
    httpMethod: "POST",
    path: "indexes",
    urlParameters: [
        Parameters.endpoint
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    requestBody: {
        parameterPath: "index",
        mapper: Object.assign(Object.assign({}, Mappers.SearchIndex), { required: true })
    },
    responses: {
        201: {
            bodyMapper: Mappers.SearchIndex
        },
        default: {
            bodyMapper: Mappers.SearchError
        }
    },
    serializer
};
const listOperationSpec = {
    httpMethod: "GET",
    path: "indexes",
    urlParameters: [
        Parameters.endpoint
    ],
    queryParameters: [
        Parameters.select,
        Parameters.apiVersion
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ListIndexesResult
        },
        default: {
            bodyMapper: Mappers.SearchError
        }
    },
    serializer
};
const createOrUpdateOperationSpec = {
    httpMethod: "PUT",
    path: "indexes('{indexName}')",
    urlParameters: [
        Parameters.endpoint,
        Parameters.indexName
    ],
    queryParameters: [
        Parameters.allowIndexDowntime,
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.ifMatch,
        Parameters.ifNoneMatch,
        Parameters.prefer
    ],
    requestBody: {
        parameterPath: "index",
        mapper: Object.assign(Object.assign({}, Mappers.SearchIndex), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.SearchIndex
        },
        201: {
            bodyMapper: Mappers.SearchIndex
        },
        default: {
            bodyMapper: Mappers.SearchError
        }
    },
    serializer
};
const deleteMethodOperationSpec = {
    httpMethod: "DELETE",
    path: "indexes('{indexName}')",
    urlParameters: [
        Parameters.endpoint,
        Parameters.indexName
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.ifMatch,
        Parameters.ifNoneMatch
    ],
    responses: {
        204: {},
        404: {},
        default: {
            bodyMapper: Mappers.SearchError
        }
    },
    serializer
};
const getOperationSpec = {
    httpMethod: "GET",
    path: "indexes('{indexName}')",
    urlParameters: [
        Parameters.endpoint,
        Parameters.indexName
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    responses: {
        200: {
            bodyMapper: Mappers.SearchIndex
        },
        default: {
            bodyMapper: Mappers.SearchError
        }
    },
    serializer
};
const getStatisticsOperationSpec = {
    httpMethod: "GET",
    path: "indexes('{indexName}')/search.stats",
    urlParameters: [
        Parameters.endpoint,
        Parameters.indexName
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    responses: {
        200: {
            bodyMapper: Mappers.GetIndexStatisticsResult
        },
        default: {
            bodyMapper: Mappers.SearchError
        }
    },
    serializer
};
const analyzeOperationSpec = {
    httpMethod: "POST",
    path: "indexes('{indexName}')/search.analyze",
    urlParameters: [
        Parameters.endpoint,
        Parameters.indexName
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    requestBody: {
        parameterPath: "request",
        mapper: Object.assign(Object.assign({}, Mappers.AnalyzeRequest), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.AnalyzeResult
        },
        default: {
            bodyMapper: Mappers.SearchError
        }
    },
    serializer
};
