// @azure/search-documents/dist-esm/src/generated/service/operations/synonymMaps.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as coreHttp from "@azure/core-http";
import * as Mappers from "../models/synonymMapsMappers";
import * as Parameters from "../models/parameters";
/** Class representing a SynonymMaps. */
export class SynonymMaps {
    /**
     * Create a SynonymMaps.
     * @param {SearchServiceClientContext} client Reference to the service client.
     */
    constructor(client) {
        this.client = client;
    }
    createOrUpdate(synonymMapName, synonymMap, options, callback) {
        return this.client.sendOperationRequest({
            synonymMapName,
            synonymMap,
            options
        }, createOrUpdateOperationSpec, callback);
    }
    deleteMethod(synonymMapName, options, callback) {
        return this.client.sendOperationRequest({
            synonymMapName,
            options
        }, deleteMethodOperationSpec, callback);
    }
    get(synonymMapName, options, callback) {
        return this.client.sendOperationRequest({
            synonymMapName,
            options
        }, getOperationSpec, callback);
    }
    list(options, callback) {
        return this.client.sendOperationRequest({
            options
        }, listOperationSpec, callback);
    }
    create(synonymMap, options, callback) {
        return this.client.sendOperationRequest({
            synonymMap,
            options
        }, createOperationSpec, callback);
    }
}
// Operation Specifications
const serializer = new coreHttp.Serializer(Mappers);
const createOrUpdateOperationSpec = {
    httpMethod: "PUT",
    path: "synonymmaps('{synonymMapName}')",
    urlParameters: [
        Parameters.endpoint,
        Parameters.synonymMapName
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.ifMatch,
        Parameters.ifNoneMatch,
        Parameters.prefer
    ],
    requestBody: {
        parameterPath: "synonymMap",
        mapper: Object.assign(Object.assign({}, Mappers.SynonymMap), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.SynonymMap
        },
        201: {
            bodyMapper: Mappers.SynonymMap
        },
        default: {
            bodyMapper: Mappers.SearchError
        }
    },
    serializer
};
const deleteMethodOperationSpec = {
    httpMethod: "DELETE",
    path: "synonymmaps('{synonymMapName}')",
    urlParameters: [
        Parameters.endpoint,
        Parameters.synonymMapName
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.ifMatch,
        Parameters.ifNoneMatch
    ],
    responses: {
        204: {},
        404: {},
        default: {
            bodyMapper: Mappers.SearchError
        }
    },
    serializer
};
const getOperationSpec = {
    httpMethod: "GET",
    path: "synonymmaps('{synonymMapName}')",
    urlParameters: [
        Parameters.endpoint,
        Parameters.synonymMapName
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    responses: {
        200: {
            bodyMapper: Mappers.SynonymMap
        },
        default: {
            bodyMapper: Mappers.SearchError
        }
    },
    serializer
};
const listOperationSpec = {
    httpMethod: "GET",
    path: "synonymmaps",
    urlParameters: [
        Parameters.endpoint
    ],
    queryParameters: [
        Parameters.select,
        Parameters.apiVersion
    ],
    responses: {
        200: {
            bodyMapper: Mappers.ListSynonymMapsResult
        },
        default: {
            bodyMapper: Mappers.SearchError
        }
    },
    serializer
};
const createOperationSpec = {
    httpMethod: "POST",
    path: "synonymmaps",
    urlParameters: [
        Parameters.endpoint
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    requestBody: {
        parameterPath: "synonymMap",
        mapper: Object.assign(Object.assign({}, Mappers.SynonymMap), { required: true })
    },
    responses: {
        201: {
            bodyMapper: Mappers.SynonymMap
        },
        default: {
            bodyMapper: Mappers.SearchError
        }
    },
    serializer
};
